what4/src/What4/Protocol/VerilogWriter/Backend.hs

```diff
@@ -0,0 +1,398 @@
+{-
+Module           : What4.Protocol.VerilogWriter.Backend
+Copyright        : (c) Galois, Inc 2020
+Maintainer       : Jennifer Paykin <jpaykin@galois.com>
+License          : BSD3
+
+Convert What4 expressions into the data types defined in the @What4.Protocol.VerilogWriter.AST@ module.
+-}
+{-# LANGUAGE GADTs, GeneralizedNewtypeDeriving, ScopedTypeVariables, RankNTypes,
+  TypeApplications, PolyKinds, DataKinds, ExplicitNamespaces, TypeOperators,
+  LambdaCase, FlexibleContexts, LambdaCase, OverloadedStrings #-}
+
+module What4.Protocol.VerilogWriter.Backend
+  where
+
+
+import           Control.Monad (mapM, foldM)
+import           Control.Monad.State (get)
+import           Control.Monad.Except
+import qualified Data.BitVector.Sized as BV
+import           Data.List.NonEmpty ( NonEmpty(..) )
+
+import           Data.Parameterized.Context
+import           GHC.TypeNats
+
+
+import qualified What4.Expr.BoolMap as BMap
+import           What4.BaseTypes as WT
+import           What4.Expr.Builder
+import           What4.Interface
+import qualified What4.SemiRing as SR
+import qualified What4.Expr.WeightedSum as WS
+import qualified What4.Expr.UnaryBV as UBV
+
+import What4.Protocol.VerilogWriter.AST
+
+doNotSupportError :: MonadError String m => String -> m a
+doNotSupportError cstr = throwError $ doNotSupportMsg ++ cstr
+
+doNotSupportMsg :: String
+doNotSupportMsg = "the Verilog backend to What4 does not support "
+
+-- | Convert What4 expresssions into verilog expressions
+exprToVerilogExpr ::
+  (IsExprBuilder sym, SymExpr sym ~ Expr n) =>
+  Expr n tp ->
+  VerilogM sym n (IExp tp)
+exprToVerilogExpr e = do
+  cache <- vsExpCache <$> get
+  let cacheEval go = idxCacheEval cache e (go e)
+  cacheEval $
+    \case
+      SemiRingLiteral (SR.SemiRingBVRepr _ w) i _ ->
+        litBV w i
+      SemiRingLiteral _ _ _ ->
+        doNotSupportError "non-bit-vector literals"
+      BoolExpr b _   -> litBool b
+      StringExpr _ _ -> doNotSupportError "strings"
+      AppExpr app -> appExprVerilogExpr app
+      NonceAppExpr n -> nonceAppExprVerilogExpr n
+      BoundVarExpr x ->
+        do name <- addFreshInput tp base
+           return $ Ident tp name
+        where
+          tp = bvarType x
+          base = bvarIdentifier x
+
+eqToVerilogExpr ::
+  (IsExprBuilder sym, SymExpr sym ~ Expr n) =>
+  Expr n tp ->
+  Expr n tp ->
+  VerilogM sym n (IExp WT.BaseBoolType)
+eqToVerilogExpr x y = do
```

Comment:
This is going to create two separate expression caches and potentially duplicate work.  It is also going to create separate copies of inputs for the same bound variables, which is probably incorrect.

Reply:
This code actually isn't used anymore (which I guess is why I never saw that blowup happen), so perhaps I'll just delete it.